File: sdks/dotnet/src/Models/EvaluationContext.cs
using System.Text.Json.Serialization;

namespace PhaseFlag.Models;

/// <summary>
/// User/request attributes sent to targeting-rule evaluation.
/// <para>
/// <see cref="UserId"/> is the primary identifier for percentage rollouts;
/// <see cref="SessionId"/> is a fallback when <see cref="UserId"/> is unavailable.
/// </para>
/// </summary>
public sealed class EvaluationContext
{
    /// <summary>Stable user identifier used for percentage rollouts.</summary>
    [JsonPropertyName("user_id")]
    public string? UserId { get; set; }

    /// <summary>Fallback identifier when <see cref="UserId"/> is unavailable.</summary>
    [JsonPropertyName("session_id")]
    public string? SessionId { get; set; }

    /// <summary>Arbitrary key-value pairs checked by targeting conditions.</summary>
    [JsonPropertyName("attributes")]
    public Dictionary<string, object?> Attributes { get; set; } = [];

    /// <summary>
    /// Look up a value by <paramref name="key"/>. Checks well-known fields
    /// (<c>user_id</c>, <c>session_id</c>) first, then falls back to
    /// <see cref="Attributes"/>.
    /// </summary>
    public object? Get(string key)
    {
        return key switch
        {
            "user_id" => UserId,
            "session_id" => SessionId,
            _ => Attributes.TryGetValue(key, out var value) ? value : null,
        };
    }
}

File: sdks/php/src/Models/EvaluationContext.php
<?php

declare(strict_types=1);

namespace PhaseFlag\Models;

/**
 * User/request attributes sent to targeting-rule evaluation.
 *
 * Contains a stable user identifier (for percentage rollouts), an optional
 * session identifier, and an arbitrary map of attributes that targeting
 * conditions can inspect.
 */
final class EvaluationContext
{
    /**
     * @param string|null              $userId     Stable user identifier used for percentage rollouts.
     * @param string|null              $sessionId  Fallback identifier when userId is unavailable.
     * @param array<string, mixed>     $attributes Arbitrary key-value pairs checked by targeting conditions.
     */
    public function __construct(
        public readonly ?string $userId = null,
        public readonly ?string $sessionId = null,
        public readonly array $attributes = [],
    ) {}

    /**
     * Look up a value by key -- checks standard fields first, then falls back
     * to the attributes map.
     */
    public function get(string $key, mixed $default = null): mixed
    {
        return match ($key) {
            'user_id'    => $this->userId ?? $default,
            'session_id' => $this->sessionId ?? $default,
            default      => $this->attributes[$key] ?? $default,
        };
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return [
            'user_id'    => $this->userId,
            'session_id' => $this->sessionId,
            'attributes' => $this->attributes,
        ];
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): static
    {
        return new static(
            userId: $data['user_id'] ?? null,
            sessionId: $data['session_id'] ?? null,
            attributes: $data['attributes'] ?? [],
        );
    }
}
